### PR TITLE
WIP:  Add minimal distance argument to local_maxima

### LINKED ISSUE #3816: Add minimal distance argument to local_maxima
https://github.com/scikit-image/scikit-image/pull/3809#issuecomment-475874038 got me thinking that an argument specifying the minimal allowed euclidean distance between maxima could be very useful. I imagine it to work akin to [this implementation](https://github.com/scipy/scipy/blob/715f3ad659fbe1f8e4b9412bf4433f85cf21dff5/scipy/signal/_peak_finding_utils.pyx#L90-L157) in SciPy. The difficulty for the n-dimensional case is to only evaluate close neighbors as otherwise the complexity grows very unfavorably with the number of maxima. I wonder whether a [KDTree](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.html) could help optimizing that particular problem. 

It's only an idea right now but before I investigate I'd like to know if there is an interest in adding this feature.

### PR DESCRIPTION
**Pull request by lagru**
## Description

Closes #3816. 

Adds the optional keyword `distance` to [local_maxima](https://scikit-image.org/docs/dev/api/skimage.morphology.html#skimage.morphology.local_maxima) (and local_minima) to specify the minimal euclidean distance allowed between maxima. In case of a conflict, the maximum with the smaller value is dismissed.

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by lagru** [comment]
The relevant implementation is in Cython and uses a rather "naive brute force" algorithm:
- Iterate maxima in order of priority (here maxima with higher image values are prioritized)
   - *continue* if current maximum was removed in an earlier step
   - Find all points of current maxima
   - Find all points within minimal distance of any point in the current maxima
   - Remove all other maxima that reach into this volume

Early benchmarks indicate a slow-down by 400% and more depending on how the `distance`-keyword is used. I still want to benchmark this more thoroughly and see if I can find any optimizations and how it compares to similar functions: e.g. [peak_Local_max](https://scikit-image.org/docs/dev/api/skimage.feature.html?highlight=peak_local_max#skimage.feature.peak_local_max) and [find_peaks](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html) come to mind.

The current API is not necessarily the best one. It may be useful to make this public or add this keyword to [h_maxima](https://scikit-image.org/docs/dev/api/skimage.morphology.html#skimage.morphology.h_maxima) as well.

**Comment by lagru** [comment]
Can't believe I forgot my earlier attempt https://github.com/scikit-image/scikit-image/issues/3016#issuecomment-405087610. :smile: I'll need to investigate how these two approaches compare...

**Comment by lagru** [comment]
Okay, a quick evaluation hints that the implementation in https://github.com/scikit-image/scikit-image/issues/3016#issuecomment-405087610 is significantly slower and scales very badly with an increasing number of peaks.  So the new approach seems like the way to go.

**Comment by jni** [comment]
@lagru very cool! Sorry for the silence, it's been busy. =)

I haven't had a chance to fully understand the code yet. With my review, I'll probably aim to break the function into some smaller functions so that the flow can be clearer. In the meantime though, it would be great to add some tests to this, and in particular to verify that functions where `peak_local_max` is currently doing a poor job actually work with this approach, e.g. https://github.com/scikit-image/scikit-image/issues/2728

Also, imho the keyword argument should be called `minimum_distance`, rather than just `distance`.

Thanks!

**Comment by jni** [comment]
@lagru another point, about your initial implementation: you were using brute-force to compare the distance to every other peak. Instead, a scipy.spatial.cKDTree could be used to rapidly search for nearby maxima, once you have the coordinates.

**Comment by lagru** [comment]
> I'll probably aim to break the function into some smaller functions so that the flow can be clearer. In the meantime though, it would be great to add some tests to this, and in particular to verify that functions where peak_local_max is currently doing a poor job actually work with this approach

Great suggestions. Definitely something to keep in mind/do next. I'm still tweaking the algorithm so I refrained from adding tests and compartmentalizing the code for now. 

[cKDTree](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.html) looks like it could be really helpful. I initially shied away because it works with "unraveled" indices (that can be solved though). I'm guessing that because it's written in C we could use it directly in Cython. Not sure whether the C-API is part of SciPy's official API though. I'll investigate and see if there's a performance improvement to be had.

File: skimage/morphology/_extrema_cy.pyx [hunk 1]
```diff
@@ -222,3 +226,191 @@ cdef inline void _fill_plateau(
...
+                        if flags[neighbor] == MAXIMUM:
+                            queue_push(&current_maximum, &neighbor)
+                            queue_count[neighbor] = 1
+                        else:
+                            queue_push(&to_search, &neighbor)
+                            queue_count[neighbor] = 1
+
+                # Evaluate the space within the minimal distance of the current
+                # maximum
+                while queue_pop(&to_search, &current_index):
```
Comment (lagru) [inline skimage/morphology/_extrema_cy.pyx]: As @jni suggested, a cKDTree might be a good tool to speed this while-block up.

**Comment by sciunto** [comment]
I don't resist to xref to this issue: #4048 Perhaps we can solve and gain also on peak_local_max.

**Comment by lagru** [comment]
@jni in the mailing list you mentioned this:

> The padding of a volume to avoid neighbor issues was done for convenience. It can and should be removed, and checking for out-of-bounds access is a mod (%) calculation away.

That'd be great to do however I don't see how this is only a modulo calculation away. As I understand it, one has to do a modulo call for all dimension in the correct order an evaluate the reminder each time (a % b = r, r=0 or r=b-1 -> border, else not border). But that wouldn't solve the problem because we actually care about not crossing the border in the wrong direction (evaluating an index at the border is fine in itself). So we'd have to consider the offsets as well.

That is quite a bit of overhead compared to padding with a border flag. Seems to me that we would be trading memory cost for runtime and complexity. Do you have examples or experience regarding this or (if you find the time) can you elaborate where this statement is coming from?

**Comment by JDWarner** [comment]
Lars is correct, it's not easy or trivial; if one goes back through my flood fill PR my immediate approach was to remove the padding.  That initially seemed to work, but actually let the fill alias off one side and into the opposite side of the image.  Fixing this became such a headache that I reverted to padding.

**Comment by lagru** [comment]
Closed in favor of #4165 which proposes an equally fast (if not faster) solution based on SciPy's cKDTree.

**Pull request closed without merging**